scripts/tank_cmd.py

```diff
@@ -379,35 +396,162 @@ def shotgun_cache_actions(log, pipeline_config_root, args):
         log.info("")
 
 
-def shotgun_run_action(log, install_root, pipeline_config_root, is_localized, args):
+def shotgun_run_action_auth(log, install_root, pipeline_config_root, is_localized, args):
     """
-    Executes the special shotgun run action command from inside of shotgun
+    Executes the special shotgun run action command from inside of shotgun.
+    Authenticated version. 
+    
+    This method expects the following arguments to be passed via the args param:
+    args[0]: name of the action
+    args[1]: entity type to operate on
+    args[2]: list of entity ids as a string, e.g. '1,2,3'
+    args[3]: shotgun user login requesting this command
+    args[4]: password for the user. Can be set to "" - in that case,
+             this is a hint to toolkit to try to authenticate via a
+             cached session token
+    args[5:]: reserved for future use. This method will *not* error 
+              if unexpected args are passed to it.
+    
+    :param log: Python logger
+    :param install_root: Root of the toolkit core installation
+    :param pipeline_config_root: Root of the pipeline configuration
+    :param is_localized: True if the pipeline configuration has been localized
+    :param args: list of arguments passed from Shotgun.
     """
-
     # we are talking to shotgun! First of all, make sure we switch on our html style logging
     log.handlers[0].formatter.enable_html_mode()
 
+    log.debug("Running shotgun_run_action_auth command")
+
+    action_name = args[0]
+    entity_type = args[1]
+    entity_ids_str = args[2].split(",")
+    entity_ids = [int(x) for x in entity_ids_str]
+    login = args[3]
+    password = args[4]
+
+    # now, first try to authenticate
+    dm = DefaultsManager()
+    sa = ShotgunAuthenticator(dm)
+
+    # first of all, if there is a default user defined, that takes precedence
+    # TODO: this interface call is likely to change.
```

Comment:
Heya!

What I need to do here is to basically say "if there is a Toolkit script user defined in shotgun.yml, then use that without asking the user any further questions - if not, then start doing authentication stuff". I was using the tk-core defaults manager to pull out the username - this code: 

https://github.com/shotgunsoftware/tk-core/blob/28031_toolkit_session_handover/python/tank/util/defaults_manager.py#L24

But I added the TODO because I know we discussed and specced out *not* having these defaults manager return user objects directly but instead just simple data - to keep the ownership of the construction of user objects internal to the security APIs (so it can change freely!). So I am not sure what you are planning on refactoring - happy to disucss. But we definitely need some way to get a User object that represents that Toolkit script on the Shotgun side in some form or another. Hope it makes sense JF! Happy to discuss more in depth when I am back.

Some of these changes are not final - because they are using security api stuff that still needs to be refactored round - so I cannot write the code correctly just yet because the underlying security API hasn't just settled yet. 

My idea here was that it is most likely still beneficial to get this change folded into the security branch - this will make it easier to do refactors later on - we still have a big code review pass that we need to do on the whole security branch. This way Francis can start testing all of the functionality in place as well! Just my 2 cents :)